### PR TITLE
Pinch zoom

### PR DESCRIPTION
**Pull request by iqmulator**

### CODE REVIEW

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts [hunk 1]
```diff
@@ -172,20 +247,54 @@ export class VisualMediaViewer {
...
+            }
+        } else if (event.pointerType == 'touch') {
+            window.addEventListener('pointermove', this.onTouchableMove);
+            this.image.style.touchAction = 'none';
+            this.imageViewer.style.touchAction = 'none';
+            this.overlay.style.touchAction = 'none';
+            this.points.push(event);
+            this.startImageRect = this.imageViewer.getBoundingClientRect();
+            this.prevImageRect = this.startImageRect;
+            if (this.points.length === 2) {
```
Comment (crui3er) [inline src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:274]: Can you add description in comments what these branches mean (here and in other methods)?
points.length == 2 or 1 or 0.
Can we have points of length only 2 or 1 or 0? Can it be above 2?

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts [hunk 2]
```diff
@@ -172,20 +247,54 @@ export class VisualMediaViewer {
...
+                this.startDistance = this.getDistance();
+                this.prevDistance = this.getDistance();
+            } else if (this.points.length === 1) {
+                this.startY = event.pageY;
+                this.startX = event.pageX;
+                this.prevY = this.startY;
+                this.prevX = this.startX;
+                let target = event.target as HTMLElement;
+                if (this.imageViewer.contains(target)) {
+                    setTimeout(() => {
```
Comment (crui3er) [inline src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:284]: Why do you need setTimeout here? Can you explain why you call footerHeaderToggle? And why does it require condition `this.points.length < 2 && !this.movementStart` ? I see logical expression but I can't guess why it means.

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts [hunk 3]
```diff
@@ -23,22 +47,60 @@ export class VisualMediaViewer {
...
 
     private round = (value: number) : number => {
         return Math.round(value);
     }
 
+    private logInfo(text: string) {
+        return this.blazorRef.invokeMethodAsync('LogJS', text);
+    }
+
+    private footerHeaderToggle() {
```
Comment (crui3er) [inline src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:93]: Better to rename to toggleFooterHeaderVisibility.

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts [hunk 4]
```diff
@@ -172,20 +247,54 @@ export class VisualMediaViewer {
...
+            if (imageTop < this.headerBottom || imageBottom > this.footerTop || imageLeft < 0 || imageRight > window.innerWidth) {
+                this.getImageAndMouseX(event, imageRect);
+                this.getImageAndMouseY(event, imageRect);
+                window.addEventListener('pointermove', this.onPointerMove);
+            } else {
+                preventDefaultForEvent(event);
+            }
+        } else if (event.pointerType == 'touch') {
+            window.addEventListener('pointermove', this.onTouchableMove);
+            this.image.style.touchAction = 'none';
```
Comment (crui3er) [inline src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:268]: Is it ok that you put here `touchAction` to `none` but never revert it back? What does it do?

**Review by crui3er — CHANGES_REQUESTED** [review]

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts [hunk 5]
```diff
@@ -1,17 +1,41 @@
...
     private readonly multiplier: number = 1.4;
     private startY: number = 0;
     private startX: number = 0;
     private deltaY: number = 0;
     private deltaX: number = 0;
+    private prevY: number = 0;
+    private prevX: number = 0;
+    private readonly originImageWidth: number = 0;
+    private readonly originImageHeight: number = 0;
+    private movementStart: boolean = false;
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:24]: bool fields should have must/is/can/etc. prefix

Comment (alexyakunin) [reply src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:24]: so "isMovementStarted"?

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts [hunk 6]
```diff
@@ -23,22 +47,60 @@ export class VisualMediaViewer {
...
     private round = (value: number) : number => {
         return Math.round(value);
     }
 
+    private logInfo(text: string) {
+        return this.blazorRef.invokeMethodAsync('LogJS', text);
+    }
+
+    private footerHeaderToggle() {
+        this.isShowFooterAndHeader = !this.isShowFooterAndHeader;
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:94]: isFooterAndHeaderShown?

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts [hunk 7]
```diff
@@ -132,6 +194,19 @@ export class VisualMediaViewer {
         return top;
     }
 
+    private getDistance() : number {
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:197]: See Vector2D type - it does all of this.

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts (hunk 2 above)
Comment (alexyakunin) [reply src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:284]: Yep, zero comments in code.

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts [hunk 8]
```diff
@@ -200,6 +309,173 @@ export class VisualMediaViewer {
 
     private onPointerUp = (event: PointerEvent) => {
         window.removeEventListener('pointermove', this.onPointerMove);
+        if (event.pointerType === 'touch') {
+            this.removeEvent(event);
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:313]: This kinda sucks. IMO event removal should be extremely reliable - i.e. somewhere in a common piece of logic.

Same for the subscription.

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts [hunk 9]
```diff
@@ -200,6 +309,173 @@ export class VisualMediaViewer {
...
+                this.scaleImage(scale);
+            }
+            this.prevDistance = distance;
+            this.prevImageRect = imageRect;
+
+        } else if (this.points.length === 1 && this.imageViewer.contains(event.target as HTMLElement)) {
+            let viewer = this.imageViewer;
+            let imageRect = this.startImageRect;
+            if (this.canMoveImage(imageRect)) {
+                let rect = this.image.getBoundingClientRect();
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:363]: My eyes bleed from here.

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts [hunk 10]
```diff
@@ -200,6 +309,173 @@ export class VisualMediaViewer {
...
+                        } else {
+                            viewer.style.top = this.round(imageRect.top + deltaY) + 'px';
+                        }
+                    }
+                    this.prevY = event.pageY;
+                }
+                this.prevImageRect = rect;
+            }
+        } else {
+            preventDefaultForEvent(event);
```
[the remark this answers is not in the document]
Comment (alexyakunin) [reply src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:415]: @crui3er curious how you managed to notice this - I mean, it's not a method, it's a poem.

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts [hunk 11]
```diff
@@ -200,6 +309,173 @@ export class VisualMediaViewer {
...
+        const windowHeight = window.innerHeight;
+        const viewerRect = this.imageViewer.getBoundingClientRect();
+
+        let left = imageLeft;
+        let top = imageTop;
+        let deltaX = (this.round(this.prevImageRect.width) - imageWidth) / 2;
+        let deltaY = (this.round(this.prevImageRect.height) - imageHeight) / 2;
+
+        // center x
+        if (rect.width >= windowWidth && rect.width <= this.prevImageRect.width) {
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:450]: This doesn't look right as well. Please make this code 3x shorter.

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts (hunk 9 above)
Comment (alexyakunin) [reply src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:363]: Please make this code 3-4x shorter.

**Review by alexyakunin — CHANGES_REQUESTED** [review]

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/VisualMediaViewerModal.razor [hunk 12]
```diff
@@ -56,6 +58,15 @@
     private void CloseModal()
         => Modal.Close();
 
+    [JSInvokable]
+    public void LogJS(string text)
+        => Log.LogInformation(text);
```
Comment (frolyo) [inline src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/VisualMediaViewerModal.razor:63]: wny is that required? looks like a vulnerability

**Review by frolyo — CHANGES_REQUESTED** [review]

File: src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts [hunk 13]
```diff
@@ -1,17 +1,41 @@
...
     private startY: number = 0;
     private startX: number = 0;
     private deltaY: number = 0;
     private deltaX: number = 0;
+    private prevY: number = 0;
+    private prevX: number = 0;
+    private readonly originImageWidth: number = 0;
+    private readonly originImageHeight: number = 0;
+    private movementStart: boolean = false;
+    private startDistance: number = 0;
```
Comment (frolyo) [inline src/dotnet/UI.Blazor/Components/VisualMediaViewerModal/visual-media-viewer.ts:25]: afair field initializer implies type, so I guess you can omit type here.

**Pull request merged**